gnucash/register/register-gnome/completioncell-gnome.c

```diff
@@ -0,0 +1,947 @@
+        if (value)
+        {
+            od = (OccurrenceData*)value;
+            od->count = od->count + 1;
+            if (!box->register_is_reversed)
+            {
+                od->date = date;
+                od->occurrence = box->occurrence;
+            }
+        }
+        else
+        {
+            od = g_new0 (OccurrenceData, 1);
+            od->date = date;
+            od->count = 1;
+            od->occurrence = box->occurrence;
+            g_hash_table_insert (box->item_hash, g_strdup (menustr), od);
+        }
+        box->occurrence++;
+    }
+}
+
+void
+gnc_completion_cell_set_value (CompletionCell* cell, const char* str)
+{
+    if (!cell || !str)
+
+    gnc_basic_cell_set_value (&cell->cell, str);
+}
+
+static inline void
+list_store_append (GtkListStore *store, char* string,
+                   char* markup, gint weight)
+{
+    GtkTreeIter iter;
+
+    g_return_if_fail (store);
+    g_return_if_fail (string);
+    g_return_if_fail (markup);
+
+    gtk_list_store_append (store, &iter);
+
+    gtk_list_store_set (store, &iter, TEXT_COL, string,
+                                      TEXT_MARKUP_COL, markup,
+                                      WEIGHT_COL, weight, -1);
+}
+
+static char*
+normalize_and_fold (char* utf8_string)
+{
+    char *normalized = g_utf8_normalize (utf8_string, -1, G_NORMALIZE_NFC);
+    if (!normalized)
+        return NULL;
+
+    char *folded = g_utf8_casefold (normalized, -1);
+    g_free (normalized);
+    return folded;
+}
+
+static void
+add_item (gpointer key, gpointer value, gpointer user_data)
+{
+    PopBox* box = user_data;
+    gchar *hash_entry = g_strdup (key);
+
+    if (hash_entry && *hash_entry)
+    {
+        gnc_utf8_strip_invalid_and_controls (hash_entry);
+        gchar *key_norm_fold = normalize_and_fold (hash_entry);
+        gchar *found = g_strstr_len (key_norm_fold, -1, box->newval);
+
+        if (found)
+        {
+            gchar *markup, *prefix = NULL, *match = NULL, *suffix = NULL;
+            gint prefix_len, suffix_len, offset;
+            gint occurrence_difference;
+            gint weight;
+            gboolean have_boundary = FALSE;
+            glong len = g_utf8_strlen (box->newval, -1);
+            gulong loc = g_utf8_pointer_to_offset (key_norm_fold, found);
+            OccurrenceData *od = (OccurrenceData*)value;
+
+            if (box->register_is_reversed)
+                occurrence_difference = od->occurrence + 1;
+            else
+                occurrence_difference = box->occurrence - od->occurrence;
+
+            if (loc > 0)
+                prefix = g_utf8_substring (hash_entry, 0, loc);
+            else
+                prefix = g_strdup ("");
+            prefix_len = g_utf8_strlen (prefix, -1);
+
+            match = g_utf8_substring (hash_entry, loc, loc + len);
+
+            if ((loc > 1) && g_strcmp0 (g_utf8_substring (hash_entry, loc - 1, loc), " ") == 0)
+               have_boundary = TRUE;
```

Comment:
ASCII space (U+0020) isn't the only possible code-point before the start of a word (It's not even the only [space character](https://en.wikipedia.org/wiki/Template:Whitespace_(Unicode))). Lots of punctuation marks and some symbols qualify too. Nor, in a UTF-8 string, can you do arithmetic on `char*`. Do it this way:
```
gunichar prev = g_utf8_get_char (g_utf8_prev_char (loc));
if (prev && (g_unichar_isspace(prev) || g_unichar_ispunct(prev))
    have_boundary = TRUE;
```

Reply:
I was going to use..

`gunichar prev = g_utf8_get_char (g_utf8_substring (hash_entry, loc - 1, loc));`

but changed to your suggestion.